scripts/compose-user-provided-service.sh

```diff
@@ -0,0 +1,164 @@
+  compose_cluster_data=$(curl -s -X GET "${COMPOSE_API_URL}/clusters" \
+    -H "Authorization: Bearer ${COMPOSE_API_TOKEN}" \
+    -H "Content-Type: application/json" \
+    2>/dev/null | jq -r ._embedded.clusters[0])
+
+  compose_cluster_id=$(echo "${compose_cluster_data}" | jq -r .id)
+  compose_account_id=$(echo "${compose_cluster_data}" | jq -r .account_id)
+}
+
+prepare_name() {
+  id=$(date +%s | md5)
+  deployment_name=$(echo "alpha-${TYPE}-${ORG}-${id}" | cut -c1-50)
+}
+
+provision_instance() {
+  local body
+  body=$(printf '{
+    "deployment": {
+      "name": "%s",
+      "account_id": "%s",
+      "cluster_id": "%s",
+      "type": "%s"
+    }
+  }' "${SERVICE_NAME:-$deployment_name}" "${compose_account_id}" "${compose_cluster_id}" "${TYPE}")
+
+  uri=$(curl -s -X POST "${COMPOSE_API_URL}/deployments" \
+      -H "Authorization: Bearer ${COMPOSE_API_TOKEN}" \
+      -H "Content-Type: application/json; charset=utf-8" \
+      -d "${body}" | jq -r .connection_strings.direct[0])
+
+  pattern='^(([a-z]{3,5})://)?((([^:\/]+)(:([^@\/]*))?@)?([^:\/?]+)(:([0-9]+))?)(\/[^?]*)?(\?[^#]*)?(#.*)?$'
+  [[ "$uri" =~ $pattern ]] || return 1;
+
+  uri=${BASH_REMATCH[0]}
+  uri_user=${BASH_REMATCH[5]}
+  uri_password=${BASH_REMATCH[7]}
+  uri_host=${BASH_REMATCH[8]}
+  uri_port=${BASH_REMATCH[10]}
+}
+
+create_user_provided_service() {
+  local body
+  body=$(printf '{
+    "username": "%s",
+    "password": "%s",
+    "host": "%s",
+    "port": "%s",
+    "uri": "%s"
+  }' "${uri_user}" "${uri_password}" "${uri_host}" "${uri_port}" "${uri}")
+
+  IFS=',';
+  for i in $(printf "%s" "${SPACE}")
+  do
+    info "Targeting organisation '${ORG}' and space '${i}'.\n"
+    cf target -o "${ORG}" -s "${i}"
+    cf create-user-provided-service "${SERVICE_NAME:-$deployment_name}" -p "${body}"
+    printf "\n"
+  done
+}
+
+check_params_and_environment() {
+  if [ -z "${COMPOSE_API_TOKEN:-}" ]; then
+    echo "COMPOSE_API_TOKEN must be exported"
+    usage
+    exit 1
+  fi
+  if [ -z "${TYPE:-}" ]; then
+    echo "Service type flag must be provided"
+    usage
+    exit 1
+  fi
+  if [ -z "${ORG:-}" ]; then
+    echo "Organisation flag must be provided"
+    usage
+    exit 1
+  fi
+  if [ -z "${SPACE:-}" ]; then
+    echo "Space flag must be provided"
+    usage
+    exit 1
+  fi
+  if [ -n "${SERVICE_NAME:-}" ]; then
+    if [ "${#SERVICE_NAME}" -gt 50 ]; then
+      abort "Name must be no longer than 50 characters."
+    fi
+  fi
+}
+
+while [[ $# -gt 0 ]]; do
+  key="$1"
+  shift
+  case $key in
+    -o|--org|--organisation)
+      ORG="$1"
+      shift
+    ;;
+    -s|--space)
+      SPACE="$1"
+      shift
+    ;;
+    -t|--type)
+      TYPE="$1"
+      shift
+    ;;
+    -n|--name)
+      SERVICE_NAME="$1"
+      shift
+    ;;
+    *)
+      # unknown option
+      usage
+    ;;
+  esac
+done
+
+load_colors
+check_params_and_environment
+prepare_name
+get_cluster_data
+provision_instance
+create_user_provided_service
+
+success "All done!"
```

Comment:
The one thing this script doesn't do is deprovision the instance if `create_user_provided_service` fails. I think we're fine without it if the script is only exploratory.

Reply:
You're right. It's nasty that it doesn't deprovision. I'm inclined to change that, but then it may be impossible to do that, as the instance may be still provisioning when we ask to delete it...

We already mention it in the usage, you need to be logged in as admin. Should we implement a `is user cf logged in` check instead?

Reply:
I think it's fine to have `cf` error like it did for me, as it's something we're intimately familiar with. I agree automated deprovisioning would be tricky, and it's making for a quite complicated script. 

The simplest solution I can think of is to output an error message if `create_user_provided_service` fails that tells the user to log in and deprovision the instance. 
The notice you have to setup whitelisting is quite visible, so I think something similar would work well?

Reply:
I've addressed this issue, lines 99-101. Please see if that's sufficient.